### DISCUSSION TITLE
Formal grammar definition for `ca65` assembly

### DISCUSSION THREAD

**Discussion by and3rson** [cc65/cc65#2158] [General]
I've recently started working on a code formatter for ca65 assembly that I called [`nice65`](https://github.com/and3rson/nice65).

It is rather primitive at this time (for example, it cannot yet "understand" macros, which is an entire beast on its own!)
I'm also focusing on 65(c)02 instruction set at this time.

I'm using [Lark](https://github.com/lark-parser/lark) (Python) as a parser library, but am considering an option of switching to Go for better portability & performance.

Even though I was able to make `nice65` work well with [C64 Kernal/BASIC](https://github.com/mist64/c64rom) sources and some other projects ("work well" == "prettify the code and not break anything"), it's still very far from perfect since it only covers the mostly used subset of features, and ca65 syntax has way more in its sleeve than I can imagine. Still, it's currently only around 200 SLOC and it's doing a relatively good job so far.

I was wondering if anyone is aware of an existing EBNF definition (or similar) for valid ca65 assembly files, or maybe there's some parsing code within ca65 itself that I can port to (or use as a reference for writing) a function that will be able to parse any valid ca65 assembly into an abstract syntax tree. Of course, the former would be the best option, but I might be asking for too much. The latter is also an option, but I'm afraid it's going to exceed my mental abilities and time constraints.
My current naїve attempt on a grammar definition for `nice65` can be found [here](https://github.com/and3rson/nice65/blob/76d66195b0cd2d27261555c0724ab42fcc0ae40d/nice65/app.py#L43).

Any thoughts and suggestions are welcome!

EDIT: Disclaimer: the set of formatting rules that I implemented in `nice65` are my personal preference. I'm planning to make indentation sizes, line breaks, etc all configurable. In fact, the formatting part itself is a breeze: the hardest part is getting `nice65` to first "understand" **any** valid ca65 assembly. Therefore adding more features (such as linting, warning about common pitfalls, etc) will also become possible, possibly in the scope of another tool.

**Comment by and3rson** [comment]
Turns out treating ca65 as a context-free grammar is really tricky. Consider an example:

```asm
.macro  ldax _a, _x
        lda _a
        ldx _x
.endmacro

start ldax _a, _x  ; In this case, parser can infer that "start" is a colon-less label
                   ; and ldax is a macro thanks to the operands
```

However, there are cases when it's impossible to unambiguously infer the terms:

```asm
.macro  pushall
        pha
        phx
        phy
.endmacro

start pushall  ; start - label, pusha - macro, or...
               ; start - macro, pusha - operand
               ; we humans can see that it's the former, but the parser doesn't.
```

At this point, colon-less labels become a real pain, since they make the grammar impossible to parse without ambiguities.
One possible solution is to make handling of colon-less labels optional via a CLI flag, and if it's enabled - then disable handling of macros. This way the parser can still handle old code (such as C64 sources), but also fully support macros and "normal" labels for modern code.

**Reply by and3rson** [reply]
I've added a flag that enables handling of colon-less labels ("legacy mode"), and it works great.
It's disabled by default, so `nice65` will treat any statement that doesn't end with `:` as macro call or mnemonic. But to preserve compatibility with legacy code (that contains ambiguous colon-less labels), `-l` option can be passed, which works as long as the analyzed code doesn't contain any macros (so `nice65` can still handle e. g.  C64 code).

**Comment by kugelfuhr** [comment]
> Turns out treating ca65 as a context-free grammar is really tricky. Consider an example:

There aren't many serious programming languages that are context free. This is also true for ca65.

Maybe it's possible to duplicate the logic that ca65 uses when parsing a line. The relevant source code can be found [here](https://github.com/cc65/cc65/blob/ce3bcadd24d3af05dc30f542722f6efb66382e82/src/ca65/main.c#L684).

**Comment by buckynbrocko** [comment]
I've been working on a tree-sitter grammar for ca65 for a little while. The grammar itself is a mess and definitely still a work in progress, but is nearly feature complete. Might be of use.

[tree-sitter-ca65](https://github.com/buckynbrocko/tree-sitter-ca65)
